### PR TITLE
cfdump support for Valhalla ImplicitCreation Attribute

### PR DESCRIPTION
**Pull request by theresa-m**
I also added the value class modifier for classes and implicit modifier for <vnew>.

Related to https://github.com/eclipse-openj9/openj9/issues/17234

### CODE REVIEW

**Comment by theresa-m** [comment]
Test class:
```
value class Test {
        public implicit Test();
        static void method2() {}
}
```

I compiled this using https://github.com/openjdk/valhalla/tree/lw5, this results in:
```
$javap -v Test

Classfile /Users/theresamammarella/localdev/ic-cfdump/Test.class
  Last modified Aug 17, 2023; size 264 bytes
  SHA-256 checksum 9df699419e1a80abe4d85b190a3a107581b3836d90ec55d5cc8f7fc6614a4067
  Compiled from "Test.java"
final value class Test
  minor version: 0
  major version: 66
  flags: (0x0050) ACC_FINAL, ACC_VALUE
  this_class: #1                          // Test
  super_class: #3                         // java/lang/Object
  interfaces: 0, fields: 0, methods: 2, attributes: 2
Constant pool:
   #1 = Class              #2             // Test
   #2 = Utf8               Test
   #3 = Class              #4             // java/lang/Object
   #4 = Utf8               java/lang/Object
   #5 = Utf8               method2
   #6 = Utf8               ()V
   #7 = Utf8               Code
   #8 = Utf8               LineNumberTable
   #9 = Utf8               <vnew>
  #10 = Utf8               ()LTest;
  #11 = Utf8               SourceFile
  #12 = Utf8               Test.java
  #13 = Utf8               ImplicitCreation
{
  static void method2();
    descriptor: ()V
    flags: (0x0008) ACC_STATIC
    Code:
      stack=0, locals=0, args_size=0
         0: return
      LineNumberTable:
        line 4: 0

  public static Test Test();
    descriptor: ()LTest;
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=1, args_size=0
         0: aconst_init   #1                  // class Test
         3: astore_0
         4: aload_0
         5: areturn
      LineNumberTable:
        line 2: 0
        line 1: 4
}
SourceFile: "Test.java"
ImplicitCreation:
  flags: (0x0001) ACC_DEFAULT
```

## Classfilewriter.cpp changes 
```
$~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump -t Test.class
Wrote 269 bytes to output file Test.j9class
$ ~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump Test.j9class 
value class Test extends java.lang.Object
{
    static void method2();
    public static implicit Test <vnew>();
}
```

## Rcdump.c changes:
```
~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump -t Test.class
Wrote 214 bytes to output file Test.j9class
~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump -x Test.j9class
 
ROM Size: 0x170 (368)
Class Name: Test
Superclass Name: java/lang/Object
Source File Name: Test.java
Basic Access Flags (0x50): 
  access: default final value
J9 Access Flags (0x4800000): (preverified) 
Class file version: 66.0
Instance Shape: 0xE
Intermediate Class Data (368 bytes): FFFFFFFFFFFFFFC0
Maximum Branch Count: 1
Interfaces (0):
ImplicitCreation flags: 0x1
Fields (0):
CP Shape Description:
  . C 

Methods (2):
  Name: method2
  Signature: ()V
  Access Flags (44008): (empty) (has debug info) 
  access: default static
  Extended modifiers (0): 
  Max Stack: 0
  Argument Count: 0
  Temp Count: 0

               lmap [empty]
               dmap [empty]
               smap [empty]
    0 JBgenericReturn 

  Debug Info:
    Line Number Table (1):
      Line:     3 PC:     0

    Variables (0):


  Name: <vnew>
  Signature: ()LTest;
  Access Flags (1040009): (has debug info) (allows final field writes) 
  access: public static implicit
  Extended modifiers (0): 
  Max Stack: 1
  Argument Count: 0
  Temp Count: 1

               lmap [    1] = 0 
               dmap [    1] = 0 
               smap [empty]
    0 JBaconst_init 1 Test
               lmap [    1] = 0 
               dmap [    1] = 0 
               smap [    1] = 1 
    3 JBastore0 
               lmap [    1] = 1 
               dmap [    1] = 1 
               smap [empty]
    4 JBaload0 
               lmap [    1] = 0 
               dmap [    1] = 1 
               smap [    1] = 1 
    5 JBgenericReturn 

  Debug Info:
    Line Number Table (2):
      Line:     2 PC:     0
      Line:     1 PC:     4

    Variables (0):


  Annotation Info:
    Field Annotations:

    Method Annotations:
```

## Main.c changes:
```
~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump -d Test.class
 
Magic: CAFEBABE
Version: 66.0
Constant Pool Size: 14
Access Flags: 0x50 ( 
  access: default final value )
ThisClass: 1 -> Test
SuperClass: 3 -> java/lang/Object
Interfaces (0):

Fields (0):
Methods (2):
  Name: 5 -> method2
  Signature: 6 -> ()V
  Access Flags: 0x8 ( 
  access: default static )
  Attributes (1):
    Code:
      Max stack: 0
      Max Locals: 0
      Code Length: 1
      Caught Exceptions (0):
      Attributes (1):
        LineNumberTable:
          pc: 0 line number: 3

  Name: 9 -> <vnew>
  Signature: 10 -> ()LTest;
  Access Flags: 0x9 ( 
  access: public static implicit  )
  Attributes (1):
    Code:
      Max stack: 1
      Max Locals: 1
      Code Length: 6
      Caught Exceptions (0):
      Attributes (1):
        LineNumberTable:
          pc: 0 line number: 2
          pc: 4 line number: 1

Attributes (2):
  SourceFile:
    Source File: 12 -> Test.java
  ImplicitCreation:
    ImplicitCreation flags: 0x1
```

## Romclasswalk.c changes
 
```
~/openj9builds/openj9-openjdk-jdk.valuetypes/build/macosx-aarch64-server-release/vm/runtime/cfdump -drx Test.class
 
	<section name="implicitCreationAttributeInfo">
		<U_32 name="implicitCreationFlags">1</U_32>
	</section>
	<section name="optionalInfo">
		<J9SRP name="optionalFileNameUTF8">0x00000046</J9SRP>
		<J9SRP name="implicitCreationAttributeSRP">0xfffffff8</J9SRP>
	</section>
```

File: runtime/cfdumper/main.c [hunk 1]
```diff
@@ -516,7 +516,18 @@ static void dumpMethod(J9CfrClassFile* classfile, J9CfrMethod* method)
...
+		if(0 == strcmp((char*)nameBytes, "<vnew>")) {
+			for \(int i = 0; i < classfile->attributesCount; i++) {
+				J9CfrAttribute* attr = classfile->attributes[i];
+				if(attr->tag == CFR_ATTRIBUTE_ImplicitCreation) {
+					j9tty_printf( PORTLIB, " implicit");
+				}
+			}
+		}
+	}
+	j9tty_printf( PORTLIB, ")\n");
```
Comment (hangshao0) [inline runtime/cfdumper/main.c:530]: I think this printf can be move to the line after `j9tty_printf( PORTLIB, " implicit");`.

File: runtime/cfdumper/main.c [hunk 2]
```diff
@@ -516,7 +516,18 @@ static void dumpMethod(J9CfrClassFile* classfile, J9CfrMethod* method)
 	j9tty_printf( PORTLIB, "  Signature: %i -> %s\n", method->descriptorIndex, classfile->constantPool[method->descriptorIndex].bytes);
 	j9tty_printf( PORTLIB, "  Access Flags: 0x%X ( ", method->accessFlags);
 	printModifiers(PORTLIB, method->accessFlags, INCLUDE_INTERNAL_MODIFIERS, MODIFIERSOURCE_METHOD);
-	j9tty_printf( PORTLIB, " )\n");
+	if(classfile->accessFlags & CFR_ACC_VALUE_TYPE) {
```
Comment (hangshao0) [inline runtime/cfdumper/main.c:519]: This block can be put inside ifdef.

File: runtime/util/rcdump.c [hunk 3]
```diff
@@ -933,9 +955,14 @@ I_32 j9bcutil_dumpRomMethod( J9ROMMethod *romMethod, J9ROMClass *romClass, J9Por
 
 	j9tty_printf( PORTLIB, "  Access Flags (%X): ", romMethod->modifiers);
 	printModifiers(PORTLIB, (U_32)romMethod->modifiers, INCLUDE_INTERNAL_MODIFIERS, MODIFIERSOURCE_METHOD);
+	if (J9_ARE_ALL_BITS_SET(romClass->optionalFlags, J9_ROMCLASS_OPTINFO_IMPLICITCREATION_ATTRIBUTE)) {
+		if ((6 == J9UTF8_LENGTH(J9ROMMETHOD_NAME(romMethod))) && strcmp((char *)J9UTF8_DATA(J9ROMMETHOD_NAME(romMethod)), "<vnew>")) {
```
Comment (hangshao0) [inline runtime/util/rcdump.c:959]: You can use one of the existing macros to do the comparison: https://github.com/eclipse-openj9/openj9/blob/d1eab2b63578a71b7d951c51dc6f25f032a28c04/runtime/oti/j9.h#L177-L185

File: runtime/util/romclasswalk.c [hunk 4]
```diff
@@ -866,8 +866,25 @@ allSlotsInPreloadAttributeDo(J9ROMClass* romClass, U_32* preloadAttributePointer
 
 	callbacks->sectionCallback(romClass, preloadAttributePointer, (UDATA)cursor - (UDATA)preloadAttributePointer, "preloadAttributeInfo", userData);
 }
-#endif /* J9VM_OPT_VALHALLA_VALUE_TYPES */
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+static void
+allSlotsInImplicitCreationAttributeDo(J9ROMClass* romClass, U_32* implicitCreationAttributePointer, J9ROMClassWalkCallbacks* callbacks, void* userData) {
```
Comment (hangshao0) [inline runtime/util/romclasswalk.c:873]: `{` should be in the new line.

File: runtime/cfdumper/main.c [hunk 5]
```diff
@@ -516,7 +516,18 @@ static void dumpMethod(J9CfrClassFile* classfile, J9CfrMethod* method)
...
+		U_8* nameBytes = classfile->constantPool[method->nameIndex].bytes;
+		if(0 == strcmp((char*)nameBytes, "<vnew>")) {
```
[the remark this answers is not in the document]
Comment (theresa-m) [reply runtime/cfdumper/main.c:521]: This file has a wide variance for whether there is a space here, there seems to be more `if(` then `if (` for what its worth.

[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/cfdumper/main.c:521]: I believe those `if(` are from old code before the J9 coding standard are introduced. Part of that standard are now documented in https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#whitespace

File: runtime/cfdumper/main.c (hunk 1 above)
Comment (hangshao0) [reply runtime/cfdumper/main.c:530]: OK, I see.

File: runtime/cfdumper/main.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (theresa-m) [reply runtime/cfdumper/main.c:521]: strcmp seems to be common practice in this file (perhaps another older practice though?).

[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/cfdumper/main.c:521]: > perhaps another older practice though?

I believe so.

[the remark this answers is not in the document]
Comment (theresa-m) [reply runtime/cfdumper/main.c]: Okay. updated

**Comment by hangshao0** [comment]
Jenkins test sanity,extended winval jdknext

**Comment by hangshao0** [comment]
Jenkins compile xlinux jdknext

**Comment by theresa-m** [comment]
Looks like an infra issue
```
JVMSHRC241E Error: unable to delete shared class cache file
JVMSHRC336E Port layer error code = -100
JVMSHRC337E Platform error message: (32) The process cannot access the file because it is being used by another process.
JVMSHRC430I Failed to remove current generation of shared class cache "jenkins"
```

**Comment by hangshao0** [comment]
From https://openj9-jenkins.osuosl.org/job/Test_openjdkValhalla_j9_sanity.functional_x86-64_windows_valhalla_Personal_testList_1/9/consoleFull, the real error is:

```
18:06:46  FAILED: testNonValueTypeClassWithImplicitCreationAttribute
18:06:46  org.testng.TestException: 
18:06:46  Method ValhallaAttributeTests.testNonValueTypeClassWithImplicitCreationAttribute()[pri:0, instance:org.openj9.test.lworld.ValhallaAttributeTests@b9d93030] should have thrown an exception of type class java.lang.ClassFormatError
18:06:46  	at org.testng.internal.ExpectedExceptionsHolder.noException(ExpectedExceptionsHolder.java:80)
18:06:46  	at org.testng.internal.Invoker.considerExceptions(Invoker.java:1139)
18:06:46  	at org.testng.internal.Invoker.invokeMethod(Invoker.java:612)
18:06:46  	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716)
18:06:46  	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988)
18:06:46  	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
18:06:46  	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
18:06:46  	at org.testng.TestRunner.privateRun(TestRunner.java:648)
18:06:46  	at org.testng.TestRunner.run(TestRunner.java:505)
18:06:46  	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
18:06:46  	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
18:06:46  	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
18:06:46  	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
18:06:46  	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
18:06:46  	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
18:06:46  	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
18:06:46  	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
18:06:46  	at org.testng.TestNG.runSuites(TestNG.java:1049)
18:06:46  	at org.testng.TestNG.run(TestNG.java:1017)
18:06:46  	at org.testng.TestNG.privateMain(TestNG.java:1354)
18:06:46  	at org.testng.TestNG.main(TestNG.java:1323)
```

**Comment by hangshao0** [comment]
Jenkins test sanity,extended zlinuxval jdknext

**Comment by theresa-m** [comment]
do I need to rebase first?

**Comment by hangshao0** [comment]
> do I need to rebase first?

No, as https://github.com/eclipse-openj9/openj9/pull/17976 is already merged.

**Comment by hangshao0** [comment]
Build launched by https://github.com/eclipse-openj9/openj9/pull/17979#issuecomment-1690436035 has passed: 

- https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/4287/

I am merging this PR.

**Comment by hangshao0** [comment]
I've approved this change.

File: runtime/cfdumper/main.c [hunk 6]
```diff
@@ -516,7 +516,20 @@ static void dumpMethod(J9CfrClassFile* classfile, J9CfrMethod* method)
 	j9tty_printf( PORTLIB, "  Signature: %i -> %s\n", method->descriptorIndex, classfile->constantPool[method->descriptorIndex].bytes);
 	j9tty_printf( PORTLIB, "  Access Flags: 0x%X ( ", method->accessFlags);
 	printModifiers(PORTLIB, method->accessFlags, INCLUDE_INTERNAL_MODIFIERS, MODIFIERSOURCE_METHOD);
-	j9tty_printf( PORTLIB, " )\n");
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+	if (classfile->accessFlags & CFR_ACC_VALUE_TYPE) {
+		J9CfrConstantPoolInfo name = classfile->constantPool[method->nameIndex];
+		if (J9UTF8_LITERAL_EQUALS(name.bytes, name.slot1, "<vnew>")) {
+			for (int i = 0; i < classfile->attributesCount; i++) {
```
Comment (keithc-ca) [inline runtime/cfdumper/main.c:523]: This is not legal C code (variables cannot be declared in `for` loops); `i` is already declared on line 511.

File: runtime/cfdumper/main.c [hunk 7]
```diff
@@ -1086,6 +1111,20 @@ static void printMethod(J9CfrClassFile* classfile, J9CfrMethod* method)
 	if(method->accessFlags & CFR_ACC_NATIVE) j9tty_printf( PORTLIB, "native ");
 	if(method->accessFlags & CFR_ACC_ABSTRACT) j9tty_printf( PORTLIB, "abstract ");
 	if(method->accessFlags & CFR_ACC_STRICT) j9tty_printf( PORTLIB, "strict ");
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+	/* ImplicitCreation is triggered by an implicit constructor in a value class */
+	if (classfile->accessFlags & CFR_ACC_VALUE_TYPE) {
+		J9CfrConstantPoolInfo name = classfile->constantPool[method->nameIndex];
+		if (J9UTF8_LITERAL_EQUALS(name.bytes, name.slot1, "<vnew>")) {
+			for (int i = 0; i < classfile->attributesCount; i++) {
```
Comment (keithc-ca) [inline runtime/cfdumper/main.c:1119]: Same comment here.

**Pull request merged**